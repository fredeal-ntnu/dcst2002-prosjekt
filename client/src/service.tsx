import axios from 'axios';

axios.defaults.baseURL = 'http://localhost:3000/api/v2';

export type Question = {
  question_id: number;
  title: string;
  text: string;
  view_count: number;
  has_answer: boolean;
  user_id: string;
};

export type Tag = {
  tag_id: number;
  name: string;
};

export type Tag_Question_Relation = {
  tag_id: number;
  question_id: number;
};

export type User = {
  user_name: string;
  password: string;
};

export type Answer = {
  answer_id: number;
  text: string;
  confirmed_answer: boolean;
  question_id: number;
  score?: number;
};

export type QuestionComment = {
  question_comment_id: number;
  text: string;
  question_id: number;
};

export type AnswerComment = {
  answer_comment_id: number;
  text: string;
  answer_id: number;
};

export type Vote = {
  user_id: number;
  answer_id: number;
  vote_type: boolean;
}

class Service {
  /**
   * Get all questions.
   */

  getAllQuestions() {
    return axios.get<Question[]>('/questions').then((response) => response.data);
  }

  /**
   * Get a question by id.
   */

  getQuestion(id: number) {
    return axios.get<Question>('/questions/' + id).then((response) => response.data);
  }

  /**
   * Get top five questions.
   */

  getTopFiveQuestions() {
    return axios.get<Question[]>('/topfivequestions').then((response) => response.data);
  }

  /**
   * Get all unaswered questions.
   */
  getUnansweredQuestions() {
    return axios.get<Question[]>('/unansweredquestions').then((response) => response.data);
  }

  /**
   * Update a question.
   */

  updateQuestion(question: Question) {
    return axios.put('/questions', question).then((response) => response.data);
  }

  /**
   * Create a question.
   */

  createQuestion(title: string, text: string, user_id: number) {
    return axios.post('/questions', { title, text, user_id }).then((response) => response.data.id);
  }

  /**
   * Delete a question.
   */

  deleteQuestion(id: number) {
    console.log(typeof id);
    return axios.delete('/questions/' + id).then((response) => response.data);
  }

  /**
   * Get tags for a question.
   */

  getTags(id: number) {
    return axios.get('/questions/' + id + '/tags').then((response) => response.data);
  }

  /**
   * Get all tags.
   */

  getAllTags() {
    return axios.get('/tags').then((response) => response.data);
  }

  createTagQuestionRelation(tag_id: number, question_id: number) {
    return axios.post('/questions/' + question_id, { tag_id, question_id });
  }

  /**
   * Get a tag by name.
   */

  getQuestionsByTagId(tag: number) {
    return axios.get('/tag/' + tag + '/questions').then((response) => response.data);
  }

  //get all tags for a question by question id
  getTagsForQuestion(id: number) {
    return axios.get('/questions/' + id + '/tags').then((response) => response.data);
  }

  getAllTagQuestionsRelations() {
    return axios.get('/question/:id').then((response) => response.data);
  }

  //get a user
  getUser(user: User) {
    return axios.get<User>('/users/' + user.user_name).then((response) => response.data);
  }

  getAnswerById(id: number) {
    return axios.get('/answers/' + id).then((response) => response.data);
  }

  getAnswersForQuestion(id: number) {
    return axios.get('/questions/' + id + '/answers').then((response) => response.data);
  }

  createAnswer(text: string, question_id: number) {
    return axios
      .post('/questions/' + question_id + '/answers', { text, question_id })
      .then((response) => response.data);
  }

  // ALLE SERVICES FOR QUESTION COMMENTS

  getQuestionCommentById(id: number) {
    return axios
    .get('/comments/'+ id)
    .then((response) => response.data);
  }

  createQuestionComment(text: string, question_id: number) {
    return axios
      .post('/questions/' + question_id + '/comments', { text, question_id })
      .then((response) => response.data);
  }

updateQuestionComment(questionComment: QuestionComment) {
  return axios
  .put('/comments/', questionComment)
  .then((response) => response.data);
}

  deleteQuestionComment(id: number) {
    return axios
      .delete('/comments/' + id)
      .then((response) => response.data);
  }

  getQuestionCommentsForQuestion(id: number) {
    return axios
    .get('/questions/' + id + '/comments')
    .then((response) => response.data);
  }

  // ALLE SERVICES FOR ANSWER COMMENTS
  

  getAnswerCommentsForAnswer(id: number) {
    return axios.get('/answers/' + id + '/comments')
    .then((response) => response.data);
  }

  createAnswerComment(text: string, answer_id: number) {
    return axios
      .post('/answers/' + answer_id + '/comments', { text, answer_id })
      .then((response) => response.data);
  }

  getAnswerCommentById(id: number) {
    console.log('andre')
    return axios
    .get('/answer/comments/' + id)
    .then((response) => response.data);
  }

  updateAnswerComment(answerComment: AnswerComment) {
    return axios
    .put('/comments/', answerComment)
    .then((response) => response.data);
  }

  deleteAnswerComment(id: number) {
    return axios
      .delete('/answer/comments/' + id)
      .then((response) => response.data);
  }

  


  
  updateAnswer(answer: Answer) {
    return axios.put('/answers', answer)
    .then((response) => response.data);
  }

deleteAnswer(id: number) {
  return axios
  .delete('/answers/' + id)
  .then((response) => response.data);
}



// alle services for votes

getVotesByAnswerId(id: number) {
  return axios
  .get('/answers/' + id + '/votes')
  .then((response) => response.data);

}

createVoteForAnswer(user_id: number, answer_id: string, vote_type: boolean) {
  return axios
  .post(answer_id + '/votes', { user_id, answer_id, vote_type })
  .then((response) => response.data);
}


}

const service = new Service();
export default service;
