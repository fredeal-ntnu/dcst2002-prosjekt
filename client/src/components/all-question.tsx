import * as React from 'react';
import { Component } from 'react-simplified';
import { Card, Alert, Column, Row, Form } from '../widgets';
import service, { Question, Tag } from '../service';

export class AllQuestions extends Component {
    questions: Question[] = [];
    filter: string = 'All'; // State to manage the filter type
    search = '';
    tags: Tag[] = [];
  
    render() {
      return (
        <>   
          <Card title='All Questions'>
            <Row>
                <Column width={2}>
                <Form.Select value={this.filter} onChange={this.handleFilterChange}>
                    <option value="all">All Questions</option>
                    <option value="popular">Most Popular</option>
                    <option value="unanswered">Unanswered</option>
                </Form.Select>
                </Column>
                <Column width={10}>
                    <Form.Input
                    type='text' 
                    value={this.search} 
                    onChange={(event) => this.search = event.currentTarget.value}
                    />
                </Column>
            </Row>

               
            {/* <Row>
            {this.tags.map((tag) => (
                        <Column>
                            <Form.Label>
                                {tag.name}
                                <input type='checkbox'
                                    value={tag.tag_id}
                                    onChange={() => {'vet ikke'}}
                                />
                            
                            </Form.Label>
                        
                        </Column>
                    ))}
            </Row> */}
          </Card>
          {this.questions
                .filter((question) => (question.title.toLowerCase().includes(this.search.toLowerCase())))
                .map((question, i) => (
                    <Card key={i} title={question.title}>
                        <Row>
                            <Column width={2}>{question.text}</Column>
                        </Row>
                    </Card>
                ))
                }
        </>
      );
    }

    handleFilterChange = (event: React.ChangeEvent<HTMLSelectElement>) => {
        this.filter = event.target.value;
        this.loadQuestions(); // Call a method to load questions based on the selected filter
      };
  
    // Method to load questions based on the current filter
    loadQuestions() {
      // Add logic here to fetch and filter questions accordingly
      // This is a placeholder for whatever your service methods might be
      switch (this.filter) {
        case 'All':
          // Fetch all questions
          break;
        case 'Popular':
          // Fetch most popular questions
          break;
        case 'Unanswered':
          // Fetch unanswered questions
          break;
        default:
          // Handle default case
      }
    }
  
    // Lifecycle method to load questions when the component mounts
    mounted() {
      this.loadQuestions();

      service
        .getAllTags()
        .then((tags) => (this.tags = tags))
        .catch((error) => Alert.danger(error.message));  
      
      service
        .getAllQuestions()
        .then((questions) => (this.questions = questions))
        .catch((error) => Alert.danger(error.message));

  }
}