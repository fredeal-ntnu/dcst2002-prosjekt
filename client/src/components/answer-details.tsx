import * as React from 'react';
import { Component } from 'react-simplified';
import { Card, Row, Column, Button, Form } from '../widgets';
import service, { Answer, AnswerComment, User } from '../service';
import { createHashHistory } from 'history';

const history = createHashHistory();

export class AnswerDetails extends Component<{ match: { params: { id: number } } }> {
  answerComments: AnswerComment[] = [];
  answer: Answer = {
    answer_id: 0,
    text: '',
    confirmed_answer: 0,
    last_updated: new Date(),
    question_id: 0,
    user_id: 0,
  };
  answerComment: AnswerComment = { answer_comment_id: 0, text: '', answer_id: 0, user_id: 0 };
  user: User = { user_id: 0, google_id: '', username: '', email: '' };
  connectedUser: number = 0;

  render() {
    return (
      <>
        <Card title="Answer">
          <Row>
            <Column width={10}>{this.answer.text}</Column>
          </Row>
        </Card>
        <Card title="Comments">
          {this.answerComments.map((answerComment) => (
            <Row key={answerComment.answer_comment_id}>
              <Column width={10}>
                {answerComment.text}
                <Column>
                  <Button.Success
                    onClick={() =>
                      history.push(
                        '/questions/' +
                          this.answer.question_id +
                          '/answers/' +
                          answerComment.answer_id +
                          '/comments/' +
                          answerComment.answer_comment_id +
                          '/edit',
                      )
                    }
                  >
                    Edit
                  </Button.Success>
                </Column>
              </Column>
            </Row>
          ))}
        </Card>

        <Card title="Add Comment">
          <Form.Textarea
            placeholder="Add comment"
            type="text"
            value={this.answerComment.text}
            onChange={(event) => (this.answerComment.text = event.currentTarget.value)}
            maxLength={1000}
            style={{ width: '500px' }}
          />
          <Column>
            <Button.Success onClick={this.addAnswerComment}>Add Comment</Button.Success>
          </Column>
        </Card>
      </>
    );
  }

  mounted() {
    service

      //Get logged in user
      .getMe()
      .then((user) => {
        this.user = user;
        this.connectedUser = this.user.user_id;
      })
      .catch((error) => {
        console.error(error.message);
        history.push('/');
        alert('You must be logged in to access this page');
      });

    service
      .getAnswerById(this.props.match.params.id)
      .then((answer) => {
        this.answer = answer;
        return service.getAnswerCommentsForAnswer(this.answer.answer_id);
      })
      .then((answerComments) => (this.answerComments = answerComments))
      .catch((error: Error) => console.error('Error getting answer: ' + error.message));
  }

  addAnswerComment() {
    
    if(this.connectedUser) {
      service
      .createAnswerComment(this.answerComment.text, this.answer.answer_id, this.connectedUser)
      .then(() => this.mounted())
      .then(() => (this.answerComment.text = ''))
      .catch((error) => console.error('Error adding answer comment: ' + error.message));
    }else return
    
  }
}
