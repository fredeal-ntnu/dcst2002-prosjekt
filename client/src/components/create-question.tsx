import * as React from 'react';
import { Component } from 'react-simplified';
import { Card, Row, Column, SideMenu, Button, Alert, Form } from '../widgets';
import service, { Tag } from '../service';
import { createHashHistory } from 'history';

const history = createHashHistory(); // Use history.push(...) to programmatically change path, for instance after successfully saving a student
export class CreateQuestion extends Component {
  tags: Tag[] = [];
  selectedTags: number[] = [];
  title = '';
  text = '';
  user_id = 0;
  connectedUser: number = 0;

  render() {
    return (
      <>
        <Row>
          <Column width={3}>
            <SideMenu
              header="Menu"
              items={[
                { label: 'Questions', to: '/questions' },
                { label: 'My Questions', to: 'myquestions' },
                { label: 'My Favourite Answers', to: 'favourites' },
                { label: 'New Question', to: 'createquestion' },
                { label: 'Tags', to: '/tags' },
              ]}
            />
          </Column>
          <Card title="Ask a Question">
            <Row>
              <Column width={2}>
                <Form.Label>Title:</Form.Label>
              </Column>
              <Column>
                <Form.Input
                  id="createTitle"
                  type="text"
                  value={this.title}
                  onChange={(event) => (this.title = event.currentTarget.value)}
                  maxLength={255}
                />
              </Column>
            </Row>
            <br />
            <Row>
              <Column width={2}>
                <Form.Label>Text:</Form.Label>
              </Column>
              <Column>
                <Form.Textarea
                  id="createText"
                  placeholder="Text"
                  value={this.text}
                  onChange={(event) => (this.text = event.currentTarget.value)}
                  rows={5}
                  maxLength={1000}
                />
              </Column>
            </Row>
            <br />
            <Row>
              <Column width={2}>
                <Form.Label>Tags:</Form.Label>
              </Column>
              <Column>
                <Row>
                  {this.tags.map((tag) => (
                    <Column key={tag.tag_id}>
                      <Form.Label>
                        {tag.name}
                        <Form.Checkbox
                          type="checkbox"
                          value={tag.tag_id}
                              checked={this.selectedTags.includes(tag.tag_id)}
                          onChange={(event) => {
                            this.handleCheckboxChange(event);
                          }}
                        />
                      </Form.Label>
                    </Column>
                  ))}
                </Row>
              </Column>
            </Row>
            <br />
            <Row>
              <Column>
                <Button.Success
                  onClick={() => {
                    this.handleAddQuestion();
                  }}
                >
                  AskMorgan
                </Button.Success>
              </Column>
            </Row>
          </Card>
        </Row>
      </>
    );
  }

  mounted() {
    service
      .getMe()
      .then((user) => {
        this.user_id = user.user_id;
        this.connectedUser = this.user_id;
      })
      .catch((error) => {
        console.error(error.message);
        history.push('/');
        alert('You must be logged in to create a question');
      });

    service
      .getAllTags()
      .then((tags) => (this.tags = tags))
      .catch((error) => console.error(error.message));
  }

  handleCheckboxChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    let tagId = Number(event.target.value);

    if (event.target.checked) {
      this.selectedTags.push(tagId);
    } else {
      this.selectedTags = this.selectedTags.filter((id) => id !== tagId);
    }
  };

  handleAddQuestion = async () => {
    if (this.selectedTags.length == 0) {
      //error
      alert('You must select at least one tag');
      return null;
    } else {
      var question_id = await service.createQuestion(this.title, this.text, this.user_id);
    }

    // For each selected tag, create a new relation in the Tag_question_relation tabl
    this.selectedTags.forEach((tag_id) => {
      service.createTagQuestionRelation(tag_id, question_id);
    });

    // Redirect to the question page
    history.push(`/questions/${question_id}`);
  };
}
