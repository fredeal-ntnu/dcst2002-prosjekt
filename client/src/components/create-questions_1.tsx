import * as React from 'react';
import { Component } from 'react-simplified';
import { Alert, Form, Card, Row, Column, Button} from '.././widgets';
import { NavLink } from 'react-router-dom';
import service, { Question, Tag_Question_Relation, Tag } from '.././service';
import { createHashHistory } from 'history';

const history = createHashHistory(); // Use history.push(...) to programmatically change path, for instance after successfully saving a student
export class CreateQuestion extends Component {
    tags: Tag[] = [];
    title = "";
    text = "";
    selectedTags: number[] = []; 
  

  render() {
    return (
          <Row>
            <Form.Label>Ask a question</Form.Label>
                <Form.Input
                    type='text'
                    placeholder='Tittel'
                    value={this.title}
                    onChange={(event) => (this.title = event.currentTarget.value)}
                />
                <Form.Input
                    type='text'
                    placeholder='Text'
                    value={this.text}
                    onChange={(event) => (this.text = event.currentTarget.value)}
                />
                <Row>
                    {this.tags.map((tag) => (
                        <Column>
                            <Form.Label>
                                {tag.name}
                                <input type='checkbox'
                                    value={tag.tag_id}
                                    onChange={(event) => {this.handleCheckboxChange(event)}}
                                />
                            
                            </Form.Label>
                        
                        </Column>
                    ))}
                   
                </Row>
                <Button.Success onClick={() => {this.handleAddQuestion()}}>Create</Button.Success>                    
          </Row>
    );
  }

  mounted() {
        
    service
      .getAllTags()
      .then((tags) => (this.tags = tags))
      .catch((error) => Alert.danger(error.message));
      console.log(this.tags)
  
  }

  handleCheckboxChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    let tagId = Number(event.target.value);

    if (event.target.checked) {
        this.selectedTags.push(tagId);
    } else {
        this.selectedTags = this.selectedTags.filter(id => id !== tagId);
    }
}


handleAddQuestion = async () => {
    console.log(typeof(this.selectedTags))
    const question_id = await service.createQuestion(this.title, this.text);

    // For each selected tag, create a new relation in the Tag_question_relation tabl
    this.selectedTags.forEach(tag_id => {
        service.createTagQuestionRelation(tag_id, question_id);
    });
    
    // Redirect to the question page
    history.push(`/questions/${question_id}`);
}


}







/*

createquestion{
  hent ut alle tagger og display på siden i form av checkbox

  insert tittel og text fra bruker
  send til tabell
  få ut result.question id

  bruk question.id til å loope igjennom alle tagger som vi har hentet, ved bruk av tag.id som vi kan sjekke med if checked.
  for alle tags funnet med den question.id, insert question.id og tag.id i tag_question_relation


}

*/