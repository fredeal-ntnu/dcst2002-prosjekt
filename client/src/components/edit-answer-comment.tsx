import * as React from 'react';
import { Component } from 'react-simplified';
import { Card, Alert, Row, Column, Button, Form} from '../widgets';
import service, {AnswerComment, User} from '../service';
import { createHashHistory } from 'history';

const history = createHashHistory();
// let history = useHistory();
export class EditAnswerComment extends Component<{ match: { params: { id: number } } }> {
  answerComments: AnswerComment[] = [];
  answerComment: AnswerComment = {answer_comment_id: 0, text: '', answer_id: 0, user_id: 0};
  user: User = { user_id: 0, google_id: '', username: '', email: '' };
  connectedUser: number = 0;
  


  render() {
    return(
      <>
      <Card title="Edit Comment">
        <Row>
          <Column width={10}>
          <Form.Textarea
                type="text"
                value={this.answerComment.text}
                onChange={(event) => (this.answerComment.text = event.currentTarget.value)}
                rows={5}
              />
          </Column>
        </Row>
        <Row>
          <Column><Button.Success onClick={this.save}>Save</Button.Success></Column>
          <Column><Button.Danger onClick={this.delete}>Delete</Button.Danger></Column>
        </Row>
      </Card>
      </>
    )
    
  }

  mounted() {
    service.getMe().then((user) => {
      this.user = user;
      this.connectedUser = user.user_id;
    });


    service.getAnswerCommentById(this.props.match.params.id)
    .then((answerComment) => (this.answerComment = answerComment))
    .catch((error: Error) => Alert.danger('Error getting answer comment: ' + error.message));
  }

  save() {
    if(this.connectedUser == this.answerComment.user_id){
    service
      .updateAnswerComment(this.answerComment)
      .then(() => history.goBack())
      .catch((error) => Alert.danger('Error saving answer comment: ' + error.message));
  }else{
    Alert.danger('You are not the owner of this comment');
  }
}

  delete() {
    if(this.connectedUser == this.answerComment.user_id){
    service
    .deleteAnswerComment(this.answerComment.answer_comment_id)
    .then(() => history.goBack())
    .catch((error) => Alert.danger('Error deleting answer comment: ' + error.message));
  }else{
    Alert.danger('You are not the owner of this comment');
  }
}

}
