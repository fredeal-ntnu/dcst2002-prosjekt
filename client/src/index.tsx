import * as React from 'react';
import { createRoot } from 'react-dom/client';
import { Component } from 'react-simplified';
import { HashRouter, Route } from 'react-router-dom';
import { NavBar, Card, Alert, CardQuestions, CardHome, Row, Column } from './widgets';
import service, { Question } from './service';
// import { TaskList, TaskDetails, TaskEdit, TaskNew } from './task-components';


class Menu extends Component {
  render() {
    return (
      <NavBar brand="askMorgan">
        <NavBar.Search/>
        <NavBar.Link to="/login">Log in</NavBar.Link>
        <NavBar.Link to="/signup">Sign up</NavBar.Link>
      </NavBar>
    );
  }
}

class Home extends Component {
  questions: Question[] = [];

  render() {
    return (
      <Card title="">
        <div className="row">
          <CardHome header="Public"
            items={[{ label: "Questions", to: "/questions" }, { label: "Tags", to: "/tags" }]}/>
          <CardQuestions header="Top Questions">
          {this.questions.map((question) => (
            <Row key={question.question_id}>
              {question.title}<br></br>{question.text}
            </Row>
          ))}
        </CardQuestions>
          
        </div> 
      </Card>
    );
  }


  createQuestion() {
    service.createQuestion().then((questions) => (this.questions = questions));
  }



    mounted() {
      service.getTopFiveQuestions().then((questions) => (this.questions = questions));
    }
}


let root = document.getElementById('root');
if (root)
  createRoot(root).render(
    <HashRouter>
      <div>
        <Alert />
        <Menu />
        <Route exact path="/" component={Home} />
        {/* <Route exact path="/signin" component={login} /> */}
        {/* <Route exact path="/signup" component={signup} /> */}
        {/* <Route exact path="/signup" component={questions} /> */}
        {/* <Route exact path="/signup" component={tags} /> */}
      </div>
    </HashRouter>,
  );
