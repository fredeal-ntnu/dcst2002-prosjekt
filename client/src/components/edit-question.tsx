import * as React from 'react';
import { createRoot } from 'react-dom/client';
import { Component } from 'react-simplified';
import { HashRouter, Route } from 'react-router-dom';
import { Card, Row, Column, SideMenu, Button, MainCard, Alert, Form } from '../widgets';
import service, { Question, Tag, Tag_Question_Relation } from '../service';
import { createHashHistory } from 'history';

const history = createHashHistory()

export class EditQuestion extends Component <{match: {params: {id: number}}}> {
    question: Question = {question_id: 0, title: "", text: "", view_count: 0, confirmed_answer: false, user_name: ""};
    
   

    render() {
        return(
            <>
            <Row>
            <Card title="Edit Question">
                <Row>
                    <Column>
                    <Form.Input
                      placeholder="Title"
                      type="text"
                      value={this.question.title}
                      onChange={(event) => (this.question.title = event.currentTarget.value)}
                    />
                    </Column>
                    <Column>
                    <Form.Textarea 
                      placeholder='Text'
                      type="text"
                      value={this.question.text}
                      onChange={(event) => (this.question.text = event.currentTarget.value)}
                      rows={5}
                      />
                    </Column>
                </Row>
            </Card> 
            </Row>
            <Row>
                <Column width={6}>
                    <Button.Success onClick={this.save}>Save</Button.Success> 
                </Column>
                <Column width={6}>
                    <Button.Danger onClick={this.delete}>Delete</Button.Danger>
                </Column>
            </Row>
            </>
            )       
       
    }

    mounted() {
        service
      .getQuestion(this.props.match.params.id)
      .then((question) => (this.question = question))
      .catch((error: Error) => Alert.danger('Error getting question: ' + error.message))
    }

    save() {
        service
          .updateQuestion(this.question)
          .then(() => history.push('/questions/' + this.question.question_id))
          .catch((error) => Alert.danger('Error saving question: ' + error.message));
      }

    delete() {
        service
          .deleteQuestion(this.props.match.params.id)
          .then(() => history.push('/questions'))
          .catch((error) => Alert.danger('Error deleting question: ' + error.message));
    }
    
}