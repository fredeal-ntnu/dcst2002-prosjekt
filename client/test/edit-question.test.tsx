import React from 'react';
import { shallow } from 'enzyme';
import { EditQuestion } from 'src/components/edit-question'; // Adjust the import path as needed
import { Column, Button } from '../src/widgets';
import { Question } from 'src/service';
import { createHashHistory } from 'history';

const history = createHashHistory();
jest.mock('../src/service',()=>{
    class Service{
        getQuestion(id: number){
            return new Promise((resolve,reject)=>{
                resolve([{question_id:1,title:'test',text:'test',view_count:1,has_answer:1,user_id:1}]);
            });
        }
        
        updateQuestion(question: Question){
            return new Promise((resolve,reject)=>{
                resolve([{question_id:1,title:'updatedtest',text:'updatedtest',view_count:1,has_answer:1,user_id:1}]);
            });
        }   
        deleteQuestion(id: number){
            return new Promise((resolve,reject)=>{
                resolve([{question_id:1,title:'test',text:'test',view_count:1,has_answer:1,user_id:1}]);
            });
        }
        
        getMe(){
            return new Promise((resolve,reject)=>{
                resolve([{user_id:1,google_id:'test',username:'test',email:'test'}]);
            });
        }
    }
    return new Service();
});

describe('Site renders', () => {
    test('Site renders correct', (done) => {
        const wrapper = shallow(<EditQuestion match={{ params: { id: 1 } }} />);
        setTimeout(()=>{
            expect(wrapper).toMatchSnapshot();
            done()
        })
    });
});

describe('Edit existing question comment', () => {
    test('edit existing question comment', (done) => {
        const wrapper = shallow(<EditQuestion match={{ params: { id: 1 } }} />);
        setTimeout(()=>{
            wrapper.setState({questionComment:{question_comment_id:0,text:'test',question_id:1,user_id:1}});
            wrapper.instance().save();
            done()
        })
    })
})

describe('Delete existing question comment', () => {
    test('delete existing question comment', (done) => {
        const wrapper = shallow(<EditQuestion match={{ params: { id: 1 } }} />);
        setTimeout(()=>{
            wrapper.setState({questionComment:{question_comment_id:0,text:'test',question_id:1,user_id:1}});
            wrapper.instance().delete();
            done()
        })
    })
})

describe('site functionality', () => {
    test('title edit field takes input', (done) => {
        const wrapper = shallow(<EditQuestion match={{ params: { id: 1 } }} />);
            wrapper.find('FormInput').simulate('change',{currentTarget:{value:'test'}});
            expect(wrapper).toMatchSnapshot();
            done()
        })
    
    test('text edit field takes input', (done) => {
        const wrapper = shallow(<EditQuestion match={{ params: { id: 1 } }} />);
            wrapper.find('FormTextarea').simulate('change',{currentTarget:{value:'test'}});
            expect(wrapper).toMatchSnapshot();
            done()
        })
    test('save button registers clcik', () => {
        let buttonClicked = false
        const wrapper = shallow(
            <Button.Success onClick={()=> (buttonClicked = true)}>test</Button.Success>,
        );
        wrapper.find('button').simulate('click');
        expect(buttonClicked).toEqual(true);
    })

    test('delete button registers click', () => {
        let buttonClicked = false
        const wrapper = shallow(
            <Button.Danger onClick={()=> (buttonClicked = true)}>test</Button.Danger>,
        );
        wrapper.find('button').simulate('click');
        expect(buttonClicked).toEqual(true);
    })
    })
