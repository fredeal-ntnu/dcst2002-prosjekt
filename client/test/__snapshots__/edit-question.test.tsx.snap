// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Site renders Site renders correct 1`] = `
<Fragment>
  <Row>
    <Card
      title="Edit Question"
    >
      <Row>
        <Column>
          <FormInput
            onChange={[Function]}
            placeholder="Title"
            type="text"
          />
        </Column>
        <Column>
          <FormTextarea
            onChange={[Function]}
            placeholder="Text"
            rows={5}
            type="text"
          />
        </Column>
      </Row>
    </Card>
  </Row>
  <Row>
    <Column
      width={6}
    >
      <ButtonSuccess
        onClick={[Function]}
      >
        Save
      </ButtonSuccess>
    </Column>
    <Column
      width={6}
    >
      <ButtonDanger
        onClick={[Function]}
      >
        Delete
      </ButtonDanger>
    </Column>
  </Row>
</Fragment>
`;

exports[`site functionality text edit field takes input 1`] = `
<Fragment>
  <Row>
    <Card
      title="Edit Question"
    >
      <Row>
        <Column>
          <FormInput
            onChange={[Function]}
            placeholder="Title"
            type="text"
            value=""
          />
        </Column>
        <Column>
          <FormTextarea
            onChange={[Function]}
            placeholder="Text"
            rows={5}
            type="text"
            value="test"
          />
        </Column>
      </Row>
    </Card>
  </Row>
  <Row>
    <Column
      width={6}
    >
      <ButtonSuccess
        onClick={[Function]}
      >
        Save
      </ButtonSuccess>
    </Column>
    <Column
      width={6}
    >
      <ButtonDanger
        onClick={[Function]}
      >
        Delete
      </ButtonDanger>
    </Column>
  </Row>
</Fragment>
`;

exports[`site functionality title edit field takes input 1`] = `
<Fragment>
  <Row>
    <Card
      title="Edit Question"
    >
      <Row>
        <Column>
          <FormInput
            onChange={[Function]}
            placeholder="Title"
            type="text"
            value="test"
          />
        </Column>
        <Column>
          <FormTextarea
            onChange={[Function]}
            placeholder="Text"
            rows={5}
            type="text"
            value=""
          />
        </Column>
      </Row>
    </Card>
  </Row>
  <Row>
    <Column
      width={6}
    >
      <ButtonSuccess
        onClick={[Function]}
      >
        Save
      </ButtonSuccess>
    </Column>
    <Column
      width={6}
    >
      <ButtonDanger
        onClick={[Function]}
      >
        Delete
      </ButtonDanger>
    </Column>
  </Row>
</Fragment>
`;
